File: uxp-plugin/js/marcaCamara.js
// Las marcas de cámara -- "[SONY] ", "[POCKET] ", "[DRONE] "-- en el
// nombre de la carpeta de un cuarto O una unidad en Premiere, y también en
// el nombre de cada clip (ver `nombre.js`).
//
// POR QUE ES UNA SEÑAL APARTE DE LA CAMARA (`label.js`): el color del
// clip ya dice de que camara salio, mirando el nombre del ARCHIVO. Esto
// mira el nombre del BIN DE IMPORTACION -- la tanda con la que Bruno
// arrastro el material -- y solo sirve para esta marca. Las dos conviven
// sin tocarse.
//
// Nacio como marcaDron.js (spec 2026-09-18, marca [DRONE] sola, regla
// todo-o-nada) y se generalizo el mismo dia a tres camaras, con la spec
// de entrega a un editor externo §4: a diferencia de [DRONE] sola --que
// se apaga ante cualquier mezcla-- un cuarto con mas de una camara se
// marca con las dos, COMBINADAS, en el orden fijo Sony/Pocket/Drone.
//
// DESDE EL 2026-09-21 LA MARCA VA AL FINAL DEL NOMBRE del cuarto
// («03. Cocina [SONY+DRONE]», no «03. [SONY+DRONE] Cocina»), para que el
// cuarto se lea primero. Ver
// docs/superpowers/specs/2026-09-21-nombre-de-clip-en-premiere-design.md.
//
// Logica pura: se prueba con `node uxp-plugin/pruebas/correr.js`.

const _MARCAS = [
  { palabra: "SONY", campo: "bin_sony" },
  { palabra: "POCKET", campo: "bin_pocket" },
  { palabra: "DRONE", campo: "bin_dron" },
];

const _MARCA_CONOCIDA = "(?:SONY|POCKET|DRONE)(?:\\+(?:SONY|POCKET|DRONE))*";

// Quita solo una marca de cámara conocida --incluida una compuesta-- y solo
// si esta al inicio O al final. Un nombre que Bruno haya escrito el mismo no
// se toca.
//
// Las DOS posiciones a proposito: la marca vive al final desde el
// 2026-09-21, pero una carpeta de una importacion anterior la trae al
// inicio. Si no se quitaran las dos, esa carpeta se leeria como otro cuarto
// y Premiere le crearia una segunda -- la familia de bug del 2026-08-22.
function sinMarcaDeCamara(nombre) {
  const s = String(nombre || "");
  return s
    .replace(new RegExp("^\\[" + _MARCA_CONOCIDA + "\\] "), "")
    .replace(new RegExp(" \\[" + _MARCA_CONOCIDA + "\\]$"), "");
}

// Las marcas que aplican a este PREFIJO de categoria_path -- un cuarto
// (["Cocina"] o ["Casa A", "Cocina"]) o una unidad sola (["Casa A"], que
// agrega la marca de TODOS los cuartos que tiene adentro).
function _marcasDelPrefijo(clipsDelManifest, prefijo) {
  return _MARCAS
    .filter(({ campo }) => _algunClipEmpiezaCon(clipsDelManifest, prefijo, campo))
    .map(({ palabra }) => palabra);
}

// La marca ya armada ("SONY", "SONY+DRONE", "SONY+POCKET+DRONE") o "" si no
// hay ninguna camara reconocible. La usan el nombre de la carpeta y el del
// clip, para que los dos digan exactamente lo mismo.
function marcaDeCamaraDelPrefijo(clipsDelManifest, prefijo) {
  const marcas = _marcasDelPrefijo(clipsDelManifest, prefijo);
  return marcas.length ? marcas.join("+") : "";
}

function _algunClipEmpiezaCon(clipsDelManifest, prefijo, campo) {
  return (clipsDelManifest || []).some((c) => {
    if (!c || !c.categoria_path || c[campo] !== true) return false;
    if (c.categoria_path.length < prefijo.length) return false;
    return prefijo.every((segmento, indice) => c.categoria_path[indice] === segmento);
  });
}

// Arma el nombre final de la carpeta: numero (si lo hay) + nombre + marca(s)
// (si aplican) -- EN ESE ORDEN, la marca al final: "03. Cocina [SONY+DRONE]".
//
// `prefijoDeCategoria` es el categoria_path (crudo, SIN numero ni marca)
// hasta este nivel: `["Cocina"]` para un cuarto sin unidad, `["Casa A",
// "Cocina"]` para un cuarto con unidad, o `["Casa A"]` para marcar la
// UNIDAD misma con la combinacion de todos sus cuartos. Si se omite, se
// asume `[nombreSinNumero]` -- el comportamiento de siempre.
function nombreDelCuartoConMarca(nombreConNumero, nombreSinNumero, clipsDelManifest, prefijoDeCategoria) {
  const prefijo = prefijoDeCategoria || [nombreSinNumero];
  const marca = marcaDeCamaraDelPrefijo(clipsDelManifest, prefijo);
  if (!marca) return nombreConNumero;
  return nombreConNumero + " [" + marca + "]";
}

File: src/clasificador_video/marca_camara.py
"""Si el nombre de un bin de importación dice de qué cámara es.

Vive aparte de `camaras.py` a propósito: ese módulo mira el nombre de los
ARCHIVOS y decide el COLOR del clip en Premiere. Este mira el nombre del
BIN DE IMPORTACIÓN -- la tanda con la que Bruno arrastró el material, la
misma que se ve como encabezado en la hoja -- y solo sirve para decidir
las marcas `[SONY]` / `[POCKET]` / `[DRONE]` en la carpeta del cuarto en
Premiere. Son dos señales distintas y no se mezclan.

Nació como `marca_dron.py` (spec 2026-09-18, la de la marca [DRONE]) y se
generalizó a las tres cámaras el mismo día, con la spec de entrega a un
editor externo, §4.
"""
from __future__ import annotations

import re


def _bin_dice(nombre: str, palabra: str) -> bool:
    return palabra in (nombre or "").lower()


def bin_dice_dron(nombre: str) -> bool:
    """`dron` es substring de `drone` («**dron**e»), así que una sola
    comparación cubre las dos formas de escribirlo."""
    return _bin_dice(nombre, "dron")


def bin_dice_sony(nombre: str) -> bool:
    return _bin_dice(nombre, "sony")


def bin_dice_pocket(nombre: str) -> bool:
    """`pocket` cubre "Osmo Pocket" completo sin tener que reconocer
    "osmo" -- que Bruno también usa para el Action, que NO entra en este
    sistema (spec 2026-09-18 §4)."""
    return _bin_dice(nombre, "pocket")


_MARCAS = (("SONY", "bin_sony"), ("POCKET", "bin_pocket"), ("DRONE", "bin_dron"))
_MARCA_CONOCIDA = r"(?:SONY|POCKET|DRONE)(?:\+(?:SONY|POCKET|DRONE))*"
_PREFIJO_MARCA = re.compile(r"^\[" + _MARCA_CONOCIDA + r"\] ")
_SUFIJO_MARCA = re.compile(r" \[" + _MARCA_CONOCIDA + r"\]$")


def sin_marca_de_camara(nombre: str) -> str:
    """Quita una única marca conocida, no texto escrito por Bruno."""
    texto = str(nombre or "")
    return _SUFIJO_MARCA.sub("", _PREFIJO_MARCA.sub("", texto))


def _algun_clip_empieza_con(clips, prefijo, campo) -> bool:
    for clip in clips or []:
        categoria = clip.get("categoria_path") if isinstance(clip, dict) else None
        if categoria is None or not clip.get(campo) or len(categoria) < len(prefijo):
            continue
        if categoria[:len(prefijo)] == prefijo:
            return True
    return False


def marca_de_camara_del_prefijo(clips, prefijo) -> str:
    return "+".join(palabra for palabra, campo in _MARCAS if _algun_clip_empieza_con(clips, prefijo, campo))


def nombre_del_cuarto_con_marca(nombre_con_numero, nombre_sin_numero, clips, prefijo_de_categoria=None) -> str:
    marca = marca_de_camara_del_prefijo(clips, prefijo_de_categoria or [nombre_sin_numero])
    return f"{nombre_con_numero} [{marca}]" if marca else nombre_con_numero
